app/models/sheet.rb

```diff
@@ -3,16 +3,45 @@ class Sheet < ActiveRecord::Base
 
   validates :home_team, :away_team, :presence => true
 
-  after_initialize :init_boxes
+  before_create :init_boxes
+  before_create :generate_code
+  before_save :close_sheet, :if => Proc.new { |sheet| sheet.all_boxes_full? && !sheet.closed?}
+  before_save :set_team_score_arrays, :if => Proc.new { |sheet| sheet.closed? && sheet.team_score_arrays_blank? }
+
+  def set_team_score_arrays
+    self.home_team_score_row = random_score_array
+    self.away_team_score_row = random_score_array
+    self.save
+  end
+
+  def random_score_array
+    (0..9).to_a.shuffle
+  end
+
+  def all_boxes_full?
+    boxes.all? { |box| box.owner_name.present? }
+  end
+
+  def close_sheet
+    self.closed = true
+  end
+
+  def team_score_arrays_blank?
+    home_team_score_row.blank? || away_team_score_row.blank?
+  end
 
   private
 
+  def generate_code
+    self.sheet_code = ("A".."Z").to_a.shuffle.sample(4).join
```

Comment:
This limits the number of possible codes by a factor of 24 (26*25*24*23 instead of 26^4).  Not a big deal at this time but maybe add a comment in here specifying that in case we need more codes at some point ?